src/Symfony/Installer/SelfUpdateCommand.php

```diff
@@ -45,71 +55,106 @@ public function isEnabled()
+            throw new \RuntimeException('Symfony Installer update failed: the "'.$this->tempDir.'" directory used to download files temporarily could not be written');
+        }
+
+        if (false === $newInstaller = @file_get_contents($this->remoteInstallerFile)) {
+            throw new \RuntimeException('The new version of the Symfony Installer couldn\'t be downloaded from the server.');
         }
 
-        if (false === @file_get_contents($remoteFilename)) {
-            $output->writeln('<error>The new version of the Symfony Installer couldn\'t be downloaded from the server.</error>');
+        $this->fs->dumpFile($this->tempInstallerFilePath, $newInstaller, 0777 & ~umask());
+
+        return $this;
+    }
 
-            return 1;
+    private function checkNewVersionIsValid()
+    {
+        // creating a Phar instance for an existing file is not allowed
+        // when the Phar extension is in readonly mode
+        if (!ini_get('phar.readonly')) {
+            // test the phar validity
+            $phar = new \Phar($this->tempInstallerFilePath);
+
+            // free the variable to unlock the file
+            unset($phar);
         }
 
-        try {
-            $tempFilename = $tempDir.'/'.basename($localFilename, '.phar').'-temp.phar';
+        return $this;
+    }
 
-            $fs->copy($remoteFilename, $tempFilename);
-            $fs->chmod($tempFilename, 0777 & ~umask());
+    private function backupOldVersion()
+    {
+        $this->fs->rename($this->localInstallerFile, $this->previousInstallerFile);
 
-            // creating a Phar instance for an existing file is not allowed
-            // when the Phar extension is in readonly mode
-            if (!ini_get('phar.readonly')) {
-                // test the phar validity
-                $phar = new \Phar($tempFilename);
-                // free the variable to unlock the file
-                unset($phar);
-            }
+        return $this;
+    }
 
-            $fs->rename($tempFilename, $localFilename, true);
+    private function replaceOldVersionbyNewVersion()
+    {
+        $fs->rename($this->tempInstallerFilePath, $this->localInstallerFile, true);
 
-            $output->writeln('<info>Symfony Installer was successfully updated.</info>');
+        return $this;
+    }
 
-            return 0;
-        } catch (\Exception $e) {
-            if (!$e instanceof \UnexpectedValueException && !$e instanceof \PharException) {
-                throw $e;
-            }
-            $fs->remove($tempFilename);
-            $output->writeln(sprintf('<error>The downloaded file is corrupted (%s).</error>', $e->getMessage()));
-            $output->writeln('<error>Please re-run the self-update command to try again.</error>');
-
-            return 1;
-        }
+    private function rollback()
+    {
+        $this->fs->remove($this->tempInstallerFilePath);
+        $this->fs->rename($this->previousInstallerFile, $this->localInstallerFile);
```

Comment:
this renaming should be done only if the roolback is done *after* the backup of the old version. Otherwise it could rallback to an even older version.

and using ``Filesystem::rename`` works only when the parent folder is writable IIRC, meaning you would break during the rollback when the command failed early.